src/services/batch-processor.spec.js

```diff
@@ -0,0 +1,165 @@
+import sinon from 'sinon';
+import TaskStore from '../stores/task-store.js';
+import TaskProcessor from './task-processor.js';
+import BatchProcessor from './batch-processor.js';
+import BatchStore from '../stores/batch-store.js';
+import { destroyTestEnvironment, pruneTestEnvironment, setupTestEnvironment } from '../test-helper.js';
+
+describe('batch-processor', () => {
+
+  const now = new Date();
+  const sandbox = sinon.createSandbox();
+
+  let sut;
+  let ctx;
+  let result;
+  let container;
+  let taskStore;
+  let batchStore;
+  let taskProcessor;
+
+  beforeAll(async () => {
+    sandbox.useFakeTimers(now);
+    container = await setupTestEnvironment();
+    taskStore = container.get(TaskStore);
+    batchStore = container.get(BatchStore);
+    taskProcessor = container.get(TaskProcessor);
+    sut = container.get(BatchProcessor);
+  });
+
+  beforeEach(() => {
+    sandbox.stub(taskStore, 'findRandomOne');
+    sandbox.stub(batchStore, 'findOne');
+    sandbox.stub(batchStore, 'save');
+    sandbox.stub(taskProcessor, 'process');
+  });
+
+  afterAll(async () => {
+    await destroyTestEnvironment(container);
+  });
+
+  afterEach(async () => {
+    await pruneTestEnvironment(container);
+    sandbox.restore();
+  });
+
+  describe('process', () => {
+    describe('when there are no batches to process', () => {
+      beforeEach(async () => {
+        ctx = { cancellationRequested: false };
+        batchStore.findOne.resolves(null);
+
+        result = await sut.process(ctx);
+      });
+
+      it('should call batchStore.findOne', () => {
+        sinon.assert.calledWith(batchStore.findOne, { completedOn: null });
+      });
+
+      it('should not call taskStore.findRandomOne', () => {
+        sinon.assert.notCalled(taskStore.findRandomOne);
+      });
+
+      it('should not call taskProcessor.process', () => {
+        sinon.assert.notCalled(taskProcessor.process);
+      });
+
+      it('should return false', () => {
+        expect(result).toBe(false);
+      });
+    });
+
+    describe('when there are no tasks to process', () => {
+      let uncompletedBatch;
+
+      beforeEach(async () => {
+        uncompletedBatch = {};
```

Comment:
While reading this, I think I found an error in our implementation (or maybe I have a knot in my brain?): After we have found an uncomplete batch, we try to get a random task for it **without** filtering by `batchId` (`taskStore.findRandomOne({ processed: false })`, from then on we process this task in connection with the uncomplete batch (e.g. by handling down the batch params to the concrete task processor. But what if there are two uncomplete batches and the randomly fetched task does not really belong to the batch we queried before? If I'm not mistaken, than we have to filter the random sample query by `batchId` :thinking:

Reply:
I lingered on the same thought a few minutes ago :))

Reply:
Very good catch, and I think we might have mentioned this while implementing and forgot to actually do it 😬 Will correct!